crates/pallet-humanode-offences/src/mock.rs

```diff
@@ -0,0 +1,312 @@
+    type WeightInfo = ();
+}
+
+parameter_types! {
+    pub const DeauthenticationReasonOnOffenceReport: DeauthenticationReason = DeauthenticationReason::Offence;
+}
+
+impl pallet_humanode_offences::Config for Test {
+    type DeauthenticationReasonOnOffenceReport = DeauthenticationReasonOnOffenceReport;
+}
+
+mock! {
+    pub Offence {
+        pub fn offenders() -> Vec<OffenderOf<Test>>;
+    }
+}
+
+impl sp_staking::offence::Offence<OffenderOf<Test>> for MockOffence {
+    const ID: sp_staking::offence::Kind = *b"-mocked-offence-";
+    type TimeSlot = ();
+
+    fn offenders(&self) -> Vec<OffenderOf<Test>> {
+        MockOffence::offenders()
+    }
+
+    fn session_index(&self) -> sp_staking::SessionIndex {
+        panic!("should be unused in tests")
+    }
+
+    fn validator_set_count(&self) -> u32 {
+        panic!("should be unused in tests")
+    }
+
+    fn time_slot(&self) -> Self::TimeSlot {
+        panic!("should be unused in tests")
+    }
+
+    fn slash_fraction(&self, _offenders_count: u32) -> sp_runtime::Perbill {
+        panic!("should be unused in tests")
+    }
+}
+
+pub fn new_test_ext() -> sp_io::TestExternalities {
+    let genesis_config = GenesisConfig {
+        session: pallet_session::GenesisConfig {
+            keys: vec![
+                (42, 42, sp_runtime::testing::UintAuthorityId(42)),
+                // Not bootnode.
+                (1, 1, sp_runtime::testing::UintAuthorityId(1)),
+            ],
+        },
+        bootnodes: pallet_bootnodes::GenesisConfig {
+            bootnodes: BoundedVec::truncate_from(vec![42]),
+        },
+        bioauth: pallet_bioauth::GenesisConfig {
+            active_authentications: BoundedVec::try_from(vec![pallet_bioauth::Authentication {
+                public_key: 1,
+                expires_at: 1000,
+            }])
+            .unwrap(),
+            ..Default::default()
+        },
+        ..Default::default()
+    };
+
+    let storage = genesis_config.build_storage().unwrap();
+
+    // Make test externalities from the storage.
+    storage.into()
+}
+
+pub fn runtime_lock() -> std::sync::MutexGuard<'static, ()> {
+    static MOCK_RUNTIME_MUTEX: std::sync::Mutex<()> = std::sync::Mutex::new(());
+
+    // Ignore the poisoning for the tests that panic.
+    // We only care about concurrency here, not about the poisoning.
+    match MOCK_RUNTIME_MUTEX.lock() {
+        Ok(guard) => guard,
+        Err(poisoned) => poisoned.into_inner(),
+    }
+}
+
+pub trait TestExternalitiesExt {
+    fn execute_with_ext<R, E>(&mut self, execute: E) -> R
+    where
+        E: for<'e> FnOnce(&'e ()) -> R;
+}
+
+impl TestExternalitiesExt for frame_support::sp_io::TestExternalities {
+    fn execute_with_ext<R, E>(&mut self, execute: E) -> R
+    where
+        E: for<'e> FnOnce(&'e ()) -> R,
+    {
+        let guard = runtime_lock();
+        let result = self.execute_with(|| execute(&guard));
+        drop(guard);
```

Comment:
```suggestion
        let result = {
            let guard = runtime_lock();
            self.execute_with(|| execute(&guard))
        };
```

Not insisting, just to discuss idioms.

Reply:
At the current moment the same style is used as for the rest our pallets,I guess it can be done in separate PR for all our pallets in our codebase if it's needed.

Reply:
I like the idea of starting organizing a library of test/mock utilities. Will create an issue as a follow up of this conversation.

Reply:
https://github.com/humanode-network/humanode/issues/1337